pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,420 @@
+	for _, reference := range latestReferences {
+		if gc.absentOwnerCache.Has(reference.UID) {
+			glog.V(5).Infof("according to the absentOwnerCache, object %s's owner %s/%s, %s does not exist", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+			dangling = append(dangling, reference)
+			continue
+		}
+		// TODO: we need to verify the reference resource is supported by the
+		// system. If it's not a valid resource, the garbage collector should i)
+		// ignore the reference when decide if the object should be deleted, and
+		// ii) should update the object to remove such references. This is to
+		// prevent objects having references to an old resource from being
+		// deleted during a cluster upgrade.
+		fqKind := schema.FromAPIVersionAndKind(reference.APIVersion, reference.Kind)
+		client, err := gc.clientPool.ClientForGroupVersionKind(fqKind)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		resource, err := gc.apiResource(reference.APIVersion, reference.Kind, len(item.identity.Namespace) != 0)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		// TODO: It's only necessary to talk to the API server if the owner node
+		// is a "virtual" node. The local graph could lag behind the real
+		// status, but in practice, the difference is small.
+		owner, err := client.Resource(resource, item.identity.Namespace).Get(reference.Name)
+		switch {
+		case IsNotFound(err):
+			gc.absentOwnerCache.Add(reference.UID)
+			glog.V(5).Infof("object %s's owner %s/%s, %s is not found", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+			dangling = append(dangling, reference)
+			continue
+		case err != nil:
+			return nil, nil, nil, err
+		}
+
+		if owner.GetUID() != reference.UID {
+			glog.V(5).Infof("object %s's owner %s/%s, %s is not found, UID mismatch", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+			gc.absentOwnerCache.Add(reference.UID)
+			dangling = append(dangling, reference)
+			continue
+		}
+
+		ownerAccessor, err := meta.Accessor(owner)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if ownerAccessor.GetDeletionTimestamp() != nil && hasDeleteDependentsFinalizer(ownerAccessor) {
+			waitingForDependentsDeletion = append(waitingForDependentsDeletion, reference)
+		} else {
+			solid = append(solid, reference)
+		}
+	}
+	return solid, dangling, waitingForDependentsDeletion, nil
+}
+
+func (gc *GarbageCollector) generateVirtualDeleteEvent(identity objectReference) {
+	event := &event{
+		eventType: deleteEvent,
+		obj:       objectReferenceToMetadataOnlyObject(identity),
+	}
+	glog.V(5).Infof("generating virtual delete event for %s\n\n", event.obj)
+	gc.dependencyGraphBuilder.enqueueChanges(event)
+}
+
+func ownerRefsToUIDs(refs []metav1.OwnerReference) []types.UID {
+	var ret []types.UID
+	for _, ref := range refs {
+		ret = append(ret, ref.UID)
+	}
+	return ret
+}
+
+func (gc *GarbageCollector) attemptToDeleteItem(item *node) error {
```

Comment:
@lavalamp this function is the main processing logic. The input the is object added to the queue by the watcher. This function takes different actions for the object that "needs to delete dependents in foreground/background", "needs to orphan dependents", "still has valid owner", or "is waiting for dependents to be deleted".